File: 프로그래머스/lv2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int solution(int k, vector<vector<int>> d) {
    int answer = -1;

    vector<int> idx(d.size());

    for (int i = 0; i < d.size(); ++i) {
        idx[i] = i;
    }
    
    do {
        vector<vector<int>> p;
        for (int i = 0; i < d.size(); ++i) {
            p.push_back(d[idx[i]]);
        }
        int fatigue = k, cnt = 0;
        for (int i = 0; i < p.size(); i++) {
            if (fatigue >= p[i][0]) {
                cnt++;
                fatigue -= p[i][1];
            }
        }
        answer = max(answer, cnt);
    } while (next_permutation(idx.begin(), idx.end()));

    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {

    static int[] arr, result;
    static boolean[] visited;
    static int len;
    static int max = Integer.MIN_VALUE;
    
    public int solution(int k, int[][] dungeons) {
        int answer = -1;
        
        len = dungeons.length;
        arr = new int[len];
        result = new int[len];
        visited = new boolean[len];
        
        for (int i = 0; i < len; i++) {
            arr[i] = i;
        }
        
        perm(0, dungeons, k);

        return max;
    }
    
    private static void perm(int depth, int[][] dungeons, int k) {
        if (depth == len) {
            go(dungeons, k);
            return;
        }
        
        for (int i = 0; i < len; i++) {
            if (!visited[i]) {
                visited[i] = true;
                result[depth] = arr[i];
                perm(depth + 1, dungeons, k);
                visited[i] = false;
            }
        }
    }
    
    private static void go(int[][] dungeons, int k) {
        int val = k;
        int cnt = 0;
        for (int i = 0; i < len; i++) {
            int need = dungeons[result[i]][0];
            int sub = dungeons[result[i]][1];
            
            if (val >= need) val -= sub;
            else break;
                
            cnt++;
        }
        
        max = Math.max(max, cnt);
    }
}